scarlet/fft.py

```diff
@@ -240,51 +241,45 @@ def __getitem__(self, index):
         # Axes that are removed from the shape of the new object
         removed = np.array([n for n, idx in enumerate(index)
                             if not isinstance(idx, slice) and idx is not None])
-        # Axes that are added to the shape of the new object
-        # (with `np.newaxis` or `None`)
-        added = np.array([n for n, idx in enumerate(index) if idx is None])
-
-        # Only propagate axes that are sliced or not indexed and
-        # decrement them by the number of removed axes smaller than each one
-        # and increment them by the number of added axes smaller than
-        # each index.
-        axes = tuple([ax-np.sum(removed < ax)+np.sum(added <= ax) for ax in self.axes if ax not in removed])
 
         # Create views into the fft transformed values, appropriately adjusting
         # the shapes for the new axes
+
         fft_kernels = {
-            tuple([s for idx, s in enumerate(shape) if self.axes[idx] not in removed]): kernel[index]
-            for shape, kernel in self._fft.items()
+            (tuple([s for idx, s in enumerate(key[0]) if key[1][idx] not in removed]),
+            tuple([a for ida, a in enumerate(key[1]) if key[1][ida] not in removed])): kernel[index]
+            for key, kernel in self._fft.items()
```

Comment:
I want to make sure that I understand the behavior of this operation. For example
```
import numpy as np
from scarlet import fft

image = fft.Fourier(np.arange(60).reshape(3, 4, 5))
image.fft((10, 20), axes=(1, 2))
image.fft((11,), axes=(1,))
image.fft((11,), axes=(2,))

print("original keys:", image._fft.keys())
print("sliced shape:", image[:, 0].shape)
print("sliced keys:", image[:, 0]._fft.keys())
```

gives

```
original keys: dict_keys([((10, 20), (1, 2)), ((11,), (1,)), ((11,), (2,))])
sliced shape: (3, 5)
sliced keys: dict_keys([((20,), (2,)), ((), ()), ((11,), (2,))])
```

So the `axes` in the shape dictionary is a `2`, not a `1`, which means the sliced `Fourier` object still holds a reference to the axes from the _original_ dimensions of the image. So it seems like the idea here is to create a `Fourier` instance of the full image, call `fft` on the unsliced image with the `axes` parameter, then slice the result. I think that we should adjust the `axes` for the sliced `Fourier` instance, otherwise if one was to ask for the `fft` of the above object they would need to supply an `axes` argument that is greater than the shape of the array, which seems confusing. It looks like you get around this by just storing the FFT of the image in `sinc_shift`, but I think that adjusting the `axes` that is stored is more general. It also includes an empty key representing the dimension that was removed, but this key should probably be removed too.

Reply:
I see what you mean, the 2 in the dictionary doesn't make sense because image[:,0] only has one axis. 
So you suggest decreasing the the value of the axis for dimensions that follow a suppressed axis right? 
My problem here is that, by slicing an Fourier object for which the transform was performed in N dimensions, we lose exactly on how many dimensions the fft was applied. Looking at the results here, I don't know if the transform was applied in one or more dimensions, but at least I know it was applied along this dimension that I have sliced. It is not ideal, but I am not sure as to how to do this exactly.

Reply:
I spent a while thinking about this problem and I don't know that there's a correct solution, since it depends on what behavior you want/expect. The best that I could come up with is to add yet another property to the `Fourier` class that keeps track of all the axes, not just the transformed ones. For example an image with dimensions `(bands, y, x)` would have `all_axes=(0, 1, 2)` and `axes=(1, 2)`. If the image is sliced, for example `images_fourier[:, 0, :]` then for the sliced `Fourier` instance `all_axes=(0, 2)`, `axes=(2,)`. I don't know if it's the best solution but at least it would keep everything consistent.

Reply:
That's also the lines along which I was thinking, though I find it a bit ugly to have an extra property and I would still find it confusing.